### PR TITLE
Add hook to register custom YARPC dispatcher

### PR DESCRIPTION
**Pull request by sectioneight**
This will be used internally to override the YARPC dispatcher.

GFM-112

### CODE REVIEW

**Comment by sectioneight** [comment]
@uber-go/service-framework

File: modules/rpc/yarpc_test.go [hunk 1]
```diff
@@ -48,6 +48,10 @@ func TestInterceptor_OK(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestRegisterDispatcher_OK(t *testing.T) {
+	RegisterDispatcher(defaultYarpcDispatcher)
```
Comment (madhuravi) [inline modules/rpc/yarpc_test.go]: assert things? Eg. Start actually invokes this function when registered.

Comment (sectioneight) [inline modules/rpc/yarpc_test.go]: That's already tested elsewhere

File: modules/rpc/yarpc.go [hunk 2]
```diff
@@ -183,3 +191,18 @@ func (m *YarpcModule) IsRunning() bool {
 	defer m.stateMu.RUnlock()
 	return m.rpc != nil
 }
+
+type yarpcDispatcherFn func(yarpc.Config) (yarpc.Dispatcher, error)
+
+// RegisterDispatcher allows you to override the YARPC dispatcher registration
+func RegisterDispatcher(dispatchFn yarpcDispatcherFn) {
+	_dispatcherFn = dispatchFn
```
Comment (glibsm) [inline modules/rpc/yarpc.go]: mutex here?

Comment (sectioneight) [inline modules/rpc/yarpc.go]: No, this should only ever be called once -- it's intentionally not thread-safe

Comment (glibsm) [inline modules/rpc/yarpc.go]: Panic on set when it's already non-nil, then? Just so we don't accidentally do it twice, or have default override the custom provided one.

Comment (sectioneight) [inline modules/rpc/yarpc.go]: It will always be non-nil, since we call Register with the default dispatcher on init

**Pull request merged**